Events for `valid` and `submit`

dhritzkiv opened the pull request.

cdaringe commented on test/callbacks.js:
```diff
@@ -84,6 +96,24 @@ test('validCallback', function(t) {
+    });
+
+    form.on('valid', function(validBool) {
+	     if (--count <= 0) {
```
indentation.  decrement in comparison works, but you could debatably more cleanly set a `t.plan(2)` at the top of your test, then put `t.end()` at the end, after `field.setValid(true);`.  Then no worrying about counting state

cdaringe commented:
i like it. +1.  did you consider if `valid` should be `session/prop`, so it just events on it's own?

dhritzkiv commented:
@cdaringe I did consider that, and started on that path. However, I realized that the event name would be `change:valid`, which could collide with events emitted for changes on a user's custom fields that may have the `name` "valid". 

Another option was making `valid` a `prop`, but under the name `_valid` (or something), but that still may be a bad idea, as user's may have code that looks similar to:
```js
form.on("all", function(eventName, field) {
    if (/^change:/.test(eventName) {
        form.model.set(field.name, field.value);
    }
})
```
as an example (I have code like this in a few places).

I feel that `change:*` events should be reserved for changes on form fields' values.

dhritzkiv commented:
(And I will fix indentation and tests later this weekend – away from computer right now)

dhritzkiv commented:
Re: using `props`/`session` for `valid`: I was mistaken about `change:` events. I for some reason thought they were only for fields changing (I only yesterday updated to ampersand-form-view@v5, so I didn't have much experience with the updated module). I now see that there are several `change` events.

That being said, what does everyone think about `valid` vs. `change:_valid`/`change:valid` for the event name? I'm still fearful of collision with fields that have the name `valid`

cdaringe commented:
At the end of the day it's probably a minor detail. change:valid gives the user more capability as it yields ~invalid detection, and a consistent interface with other & state extending protos.  No fuss, but it's my 10cents

dhritzkiv commented:
Okay. Gave this a fair bit of thought and ended up going the route of using `session` to hold `valid`, setting `valid` directly, as well as listening to the `change:valid` event. That being said, there are some big changes (methods ripped out/changed) that might need careful review/testing.

Summary:
- tests are unaltered; working as before.
- removed the `setValid` function.
    - AFAIK, this was not documented, and was only used internally within ampersand-form-view.
    - Its `forceFire` argument was only used in one place, near the end of `render` and seemed like a hack. I've mimicked it's functionality (force triggering `change:valid` on first render) by very briefly setting `valid` to `null` and then setting the proper `valid` value with `checkValid`
- removed the `forceFire` argument from `checkValid`
    - again, this was only used internally at one point
    - set `valid` directly

kamilogorek commented on ampersand-form-view.js:
```diff
@@ -24,13 +27,12 @@ module.exports = View.extend({
         this.el = opts.el;
-        this.validCallback = opts.validCallback || this.validCallback || noop;
-        this.submitCallback = opts.submitCallback || this.submitCallback || noop;
+        this.validCallback = opts.validCallback || this.validCallback || null;
```
Those `null` values are redundant, as it'll fallback to `undefined` anyway if user doesn't pass `opts` or it won't exist already.

kamilogorek commented on ampersand-form-view.js:
```diff
@@ -91,12 +93,11 @@ module.exports = View.extend({
-    checkValid: function (forceFire) {
-        var valid = this._fieldViewsArray.every(function (field) {
+    checkValid: function () {
+        this.valid = this._fieldViewsArray.every(function (field) {
```
You can instantly return assignment here `return this.valid = this.fieldViewsArray.every...`.

kamilogorek commented:
Minor comments, but other than that :+

cdaringe commented on ampersand-form-view.js:
```diff
@@ -50,6 +52,14 @@ module.exports = View.extend({
+        
+        if (this.validCallback) {
+            this.on('change:valid', function(view, validBool) {
+                this.validCallback(validBool);
```
this will call the validCallback when the form goes to invalid (validBool === false), and i believe it should only do so when true

dhritzkiv replied:
The current behaviour is that `validCallback` is called when `valid` changes between true and false. I believe this has been the case since v1.0.0.

In the docs:

```js
validCallback: function (valid) {
    if (valid) {
        console.log('The form is valid!');
    } else {
        console.log('The form is not valid!');
    }
}
```

To me this behaviour make sense.

cdaringe replied:
so you're correct. i have misspoken!  thx

cdaringe commented on ampersand-form-view.js:
```diff
@@ -50,6 +52,14 @@ module.exports = View.extend({
+            });
+        }
+
+        if (this.submitCallback) this.on('submit', this.submitCallback);
```
:+1: event listening on these form actions is a great idea

cdaringe commented on ampersand-form-view.js:
```diff
@@ -176,7 +176,8 @@ module.exports = View.extend({
         this.handleSubmit = this.handleSubmit.bind(this);
         this.el.addEventListener('submit', this.handleSubmit, false);
-        this.checkValid(true);
+        this.set('valid', null, {silent: true});
```
maybe we could annotate why the `null` setting is a requirement, so future users understand its necessity?  its not obvious on first read.

dhritzkiv replied:
Done!

wraithgar commented:
This is looking good and can merge if @cdaringe is still a +1 on it.  Really glad to see valid transition to a session variable.  Yes the `setValid` was all internal and does not count as published api.

cdaringe commented:
publish minor, ya?

dhritzkiv commented:
I'd say so. Nothing breaking in here (unless someone used the undocumented `setValid` method)

kamilogorek commented:
Agree. As @wraithgar already mentioned
> Yes the setValid was all internal and does not count as published api.
therefore it's not API breaking change.

cdaringe commented:
did a README update for this too. 4.1.0, comin in hot
